### PR TITLE
User details

### PR DESCRIPTION
**Pull request by Ruthy4**
https://github.com/Ruthy4/Random-User-Generator/assets/33810711/3937e1cb-5541-4802-a670-08a76333ad09

### CODE REVIEW

File: app/src/main/java/com/example/randomusergenerator/MainActivity.kt [hunk 1]
```diff
@@ -24,16 +36,34 @@ class MainActivity : ComponentActivity() {
...
-@Preview(showBackground = true)
-@Composable
-fun DefaultPreview() {
-    RandomUserGeneratorTheme {
-        UserScreen()
+
+    @Composable
+    fun MyApp() {
+        val navController = rememberNavController()
+        val userViewModel: UserViewModel = hiltViewModel()
```
Comment (princessdharmy) [inline app/src/main/java/com/example/randomusergenerator/MainActivity.kt:48]: Hmmmmm
I don't think you are meant to call the `UserViewModel` in the activity. What if there were several `ViewModel` classes? That would make all of their lifecycles attached to the activity which should not be rather to the view observing it.

So, this function needs to be refactored to not observe the viewmodel except the navigation. Alos, there should be a workaround to the user details.

File: app/src/main/java/com/example/randomusergenerator/user/UserDetailsScreen.kt [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        ) {
+            Image(
+                painter = rememberImagePainter(data = user.picture?.thumbnail),
+                contentDescription = stringResource(R.string.description_user_card),
+                modifier = Modifier.fillMaxSize()
+            )
+        }
+
+        Text(
+            text = "${user.name?.title} ${user.name?.first} ${user.name?.last}",
```
Comment (princessdharmy) [inline app/src/main/java/com/example/randomusergenerator/user/UserDetailsScreen.kt:64]: This could have been formatted in the data class and called directly instead of the view doing it.

File: app/src/main/java/com/example/randomusergenerator/user/UserDetailsScreen.kt [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                fontSize = fontSize20,
+                color = Color.Black
+            ),
+            modifier = Modifier.padding(top = space_x1)
+        )
+
+        Spacer(modifier = Modifier.height(space_x3_5))
+        DetailsCard(
+            stringResource(R.string.address),
+            "${user.location?.street?.name}, ${user.location?.city ?: ""}, " +
```
Comment (princessdharmy) [inline app/src/main/java/com/example/randomusergenerator/user/UserDetailsScreen.kt:85]: Same thing here. Make sure to write unit tests for any added logic to data classes.

File: app/src/main/java/com/example/randomusergenerator/user/UserScreen.kt [hunk 4]
```diff
@@ -52,7 +48,9 @@ fun UserListScreen(userViewState: UserViewState) {
             userViewState.users?.let { userList ->
                 LazyColumn(modifier = Modifier.fillMaxSize()) {
                     items(userList) { user ->
-                        UserCard(user)
+                        UserCard(user, onClick = {
+                            navController.navigate("user_details_screen/${it.id}")
```
Comment (princessdharmy) [inline app/src/main/java/com/example/randomusergenerator/user/UserScreen.kt:52]: Remember state hoisting? You could use the High Order Function instead of passing down the navcontroller.

File: app/src/main/java/com/example/randomusergenerator/user/UserScreen.kt [hunk 5]
```diff
@@ -13,25 +13,21 @@ import androidx.compose.runtime.*
...
 import com.example.randomusergenerator.user.viewmodel.UserViewModel
 
 @Composable
-fun UserScreen() {
-    val userViewModel: UserViewModel = viewModel()
+fun UserScreen(userViewModel: UserViewModel, navController: NavController) {
     val userViewState by userViewModel.uiState.collectAsState()
 
-    LaunchedEffect(key1 = true) {
-        userViewModel.getAllUsers()
```
Comment (princessdharmy) [inline app/src/main/java/com/example/randomusergenerator/user/UserScreen.kt:28]: Why did you remove this?

**Review by princessdharmy — CHANGES_REQUESTED** [review]
Please address the comments.

File: app/src/main/java/com/example/randomusergenerator/user/UserScreen.kt (hunk 5 above)
Comment (Ruthy4) [reply app/src/main/java/com/example/randomusergenerator/user/UserScreen.kt:28]: I wanted to limit the network calls made. It always called get users every time the user comes to the first screen

File: app/src/main/java/com/example/randomusergenerator/MainActivity.kt (hunk 1 above)
Comment (Ruthy4) [reply app/src/main/java/com/example/randomusergenerator/MainActivity.kt:48]: its not necessarily in the activity, its in another composable. I can extract the composable to another class

File: app/src/main/java/com/example/randomusergenerator/user/UserScreen.kt (hunk 5 above)
Comment (princessdharmy) [reply app/src/main/java/com/example/randomusergenerator/user/UserScreen.kt:28]: Kindly add a back icon to the detail page.

File: app/src/main/java/com/example/randomusergenerator/MainActivity.kt (hunk 1 above)
Comment (princessdharmy) [reply app/src/main/java/com/example/randomusergenerator/MainActivity.kt:48]: You should not reference the viewmodel outside the main caller/user. You might need to create a navigator that would do that for you. Then, navigate from your viewmodel instead from the view.
